File: 2020/python/day08.py
import unittest


def parse(line):
    op, count = line.strip().split(' ')
    count = int(count)
    return (op, count)


def step(program, idx, acc):
    (op, count) = program[idx]
    if op == 'acc':
        acc += count
        idx += 1
    elif op == 'jmp':
        idx += count
    else:
        idx += 1
    return (idx, acc)


def run(lines):
    visited = set()
    program = [parse(line) for line in lines]
    idx = 0
    acc = 0
    while True:
        if idx in visited:
            return acc
        visited.add(idx)
        (i, a) = step(program, idx, acc)
        idx = i
        acc = a


def part1():
    print(run(read_input()))


def terminate_ok(op, program, index):
    if program[index] == op:
        True
    else:
        False


def run2(lines):
    program = [parse(line) for line in lines]
    idx = 0
    acc = 0


def part2():
    pass


def read_input():
    with open('../inputs/input08') as input:
        return input.readlines()


test_input = '''nop +0
acc +1
jmp +4
acc +3
jmp -3
acc -99
acc +1
jmp -4
acc +6'''


class TestPart1(unittest.TestCase):
    def test_parse(self):
        self.assertEqual(parse(test_input.split('\n')[0]), ('nop', 0))
        self.assertEqual(parse(test_input.split('\n')[1]), ('acc', 1))
        self.assertEqual(parse(test_input.split('\n')[2]), ('jmp', 4))

    def test_step(self):
        program = [parse(line) for line in test_input.split('\n')]
        idx = 0
        acc = 0
        self.assertEqual(step(program, idx, acc), (1, 0))
        idx = 1
        acc = 0
        self.assertEqual(step(program, idx, acc), (2, 1))
        idx = 2
        acc = 1
        self.assertEqual(step(program, idx, acc), (6, 1))
        idx = 6
        acc = 1
        self.assertEqual(step(program, idx, acc), (7, 2))

    def test_run(self):
        self.assertEqual(run(test_input.split('\n')), 5)


class TestPart2(unittest.TestCase):
    def test_run2(self):
        pass


if __name__ == '__main__':
    # run(test_input.split('\n'))
    part1()
    unittest.main()

File: 2020/rust-sln/src/day08.rs
use std::collections::HashSet;
fn main() {
    println!("Part1 {}", part1());
    println!("Part2 {}", part2());
}

fn read_input() -> Vec<(&'static str, isize)> {
    include_str!("../../inputs/input08")
        .split('\n')
        .filter(|line| line.len() > 0)
        .map(|line| parse(line))
        .collect::<Vec<_>>()
}

fn parse<'a>(line: &'a str) -> (&'a str, isize) {
    let tokens = line.trim().split(' ').collect::<Vec<_>>();
    (tokens[0], tokens[1].parse::<isize>().unwrap())
}

#[derive(Debug, PartialEq, Clone)]
enum State {
    Infinite,
    Finished,
    Error,
}

fn step(instructions: &Vec<(&str, isize)>, pointer: isize, acc: isize) -> (isize, isize) {
    let (op, count) = instructions[pointer as usize];
    match op {
        "acc" => (pointer + 1, acc + count),
        "jmp" => (pointer + count, acc),
        _ => (pointer + 1, acc),
    }
}

fn out_of_bounds(pointer: isize, size: isize) -> bool {
    pointer >= size
}

fn run(instructions: &Vec<(&str, isize)>) -> (State, isize) {
    let mut visited = HashSet::new();
    let mut pointer = 0;
    let mut acc = 0;
    loop {
        if visited.get(&pointer).is_some() {
            return (State::Infinite, acc);
        }
        visited.insert(pointer);
        let (p, a) = step(instructions, pointer, acc);
        pointer = p;
        acc = a;
        if out_of_bounds(pointer, instructions.len() as isize) {
            return (State::Finished, acc);
        }
    }
}

fn part1() -> isize {
    let (state, acc) = run(&read_input());
    assert_eq!(state, State::Infinite);
    acc
}

fn replace_op(pointer: usize, instructions: &mut Vec<(&str, isize)>) {
    match instructions[pointer].0 {
        "nop" => instructions[pointer] = ("jmp", instructions[pointer].1),
        "jmp" => instructions[pointer] = ("nop", instructions[pointer].1),
        _ => (),
    };
}

fn run2(instructions: &Vec<(&str, isize)>) -> (State, isize) {
    for pointer in 0..instructions.len() {
        let mut curr_instructions = instructions.clone();
        replace_op(pointer, &mut curr_instructions);
        let (state, acc) = run(&curr_instructions);
        if state == State::Finished {
            return (state, acc);
        }
    }

    (State::Error, 0)
}

fn part2() -> isize {
    let (state, acc) = run2(&read_input());
    assert_eq!(state, State::Finished);
    acc
}

mod tests {
    use super::*;

    fn test_input() -> &'static str {
        r#"nop +0
acc +1
jmp +4
acc +3
jmp -3
acc -99
acc +1
jmp -4
acc +6"#
    }

    #[test]
    fn test_parse() {
        let input = test_input().split('\n').collect::<Vec<_>>();
        assert_eq!(parse(input[0]), ("nop", 0));
        assert_eq!(parse(input[1]), ("acc", 1));
        assert_eq!(parse(input[2]), ("jmp", 4));
    }

    #[test]
    fn test_step() {
        let instructions = test_input()
            .split('\n')
            .map(|line| parse(line))
            .collect::<Vec<_>>();
        let pointer = 0;
        let acc = 0;
        assert_eq!(step(&instructions, pointer, acc), (1, 0));
        let pointer = 1;
        let acc = 0;
        assert_eq!(step(&instructions, pointer, acc), (2, 1));
        let pointer = 2;
        let acc = 1;
        assert_eq!(step(&instructions, pointer, acc), (6, 1));
        let pointer = 6;
        let acc = 1;
        assert_eq!(step(&instructions, pointer, acc), (7, 2));
    }

    #[test]
    fn test_run() {
        let instructions = test_input()
            .split('\n')
            .map(|line| parse(line))
            .collect::<Vec<_>>();
        assert_eq!(run(&instructions), (State::Infinite, 5));
    }
}
